README.md
This module allows admins to add redirects from one URL to another within an [Apostrophe site](http://apostrophecms.org/).

This version is for [Apostrophe 2.x](http://apostrophecms.org/). For Apostrophe 0.5 use the 0.x series.

**Table of Contents**

* [Installation](#installation)

* [Configuration](#configuration)

* [Usage](#usage)

## <a id="installation"></a> Installation

First make sure you have an [Apostrophe project](http://apostrophecms.org/)!

Then:

```javascript
npm install --save apostrophe-redirects
```

## <a id="installation"></a> Configuration

In `app.js`, add the module to your configuration:

```javascript
// Other modules, then...
'apostrophe-redirects': { }
```

If you wish, you can change the default status code to `301` (permanent redirect):

```javascript
// Other modules, then...
'apostrophe-redirects': {
  statusCode: 301
}
```

> Note that permanent redirects are cached by Google for a long time. It is a good idea to encourage users to test with a temporary redirect first, then switch to permanent which is an SEO best practice — as long as it's correct.

**If you are using the workflow module with multiple locales,** we recommend turning on the new `crossLocales` option, which allows you to pick the locale for each redirect:

```javascript
'apostrophe-redirects': {
  crossLocales: true
}
```

If you are migrating existing redirects to `crossLocales: true`, you will need to run the following **one-time** command line task:

```bash
node app apostrophe-redirects:switch-to-cross-locales

This transition cannot be undone.

## <a id="usage"></a> Usage

While logged in as an admin, click the "Redirects" button. A list of redirects appears, initially empty. Add as many redirects as you like. The "from" URL must begin with a `/`. The "to" URL may be anything and need not be on your site. The "description" field is for your own convenience.

By default a redirect includes any query string (the `?` and whatever follows it, up to but not including any `#`) on incoming requests when matching for redirection. You can toggle the "ignore query string when matching" option in a redirect definition to ignore query strings on incoming requests and only match on the base URL path. A redirect that does not use this option will always match first, so you can match various specific query strings and then have a fallback rule for other cases.

Be aware that each redirect is live as soon as you save it and that it is possible to make a mess with redirects. In a pinch, you can remove unwanted redirects via the MongoDB command line client (look for `{ type: "apostrophe-redirect" }` in the `aposDocs` collection in MongoDB).

Also be aware that Apostrophe already creates "soft redirects" every time you change the slug of a page. So you shouldn't need manually created "hard redirects" in that situation.

## Advanced configuration with pieces

If you have pieces on your site and you would like to make a safer more persistent redirect that will update if your redirect destination changes, you can use polymorphic joins. To do this you need to add your piece(s) when configuring the module:

```javascript
'apostrophe-redirects': {
  withType: ['apostrophe-page', 'product']
}
```

Note: When adding your own pieces, you should always define `apostrophe-page` to still make it possible to redirect to any normal page. Also note that your piece is written in singular and that you must have a suitable apostrophe-pieces-page for your piece (that is, it must have a `._url` property when Apostrophe fetches it). Refer to the core documentation for [reusable content with pieces](https://apostrophecms.org/docs/tutorials/getting-started/reusable-content-with-pieces.html#displaying-a-directory-of-people-on-a-page-with-code-apostrophe-pieces-pages-code).

Now when you create a new redirect, you have the option to browse for both pages and your pieces in the Apostrophe admin UI.


CHANGELOG.md
# Changelog

## 2.4.2
Fixed incompatibility with sites **not** using workflow introduced in 2.4.1.

## 2.4.1
Functionality introduced in 2.4.0 was not compatible with certain older versions of MongoDB. This has been addressed for legacy reasons. There is no change in the new functionality.

## 2.4.0
In the presence of workflow and locale prefixes, a redirect created in a different locale will now work if the URL matches, falling back to the redirect target's locale even if the user's current or default locale is different.

## 2.3.2
Corrects additional defects in query string handling that led to problems with the briefly released 2.3.1.

## 2.3.1
Improves regex for "ignore query string" option to allow redirects from pages that are not slug-like such as those ending in `.html`. Also fixes a regular expression bug introduced in 2.3.0 where a rule could match if any portion of it appeared in the URL.

## 2.3.0
"Ignore query string" option for times when you'd like to match the URL regardless of the `?` and everything that follows it. Thanks to Shaun Hurley for the contribution.

## 2.2.0
Implemented polymorphic joins for internal pages which also makes it possible to configure your own pieces for polymorphic joins.

## 2.1.0
Disables `apostrophe-site-map` for redirects to prevent more superfluous UI tabs.

## 2.0.0
Implemented `statusCode` option and user-editable `statusCode` field allowing the user to choose a permanent or temporary redirect. For bc the default is still `302`. If the `statusCode` option is set to `301` instead, permanent redirects will be the default for *new* redirects. Existing redirects may be manually switched to `301` if desired.

## 0.6.6
Disables `apostrophe-seo` and `apostrophe-open-graph` for redirects to prevent superfluous UI tabs.

## 0.6.5
Use express middleware instead of relying on `apos.pages.serve`. This lets redirects happen for files, etc.

## 0.6.1
the `redirectSlug` should be `type: 'string'` because the old site may have allowed nonsense in slugs that we do not allow. Also compare to `req.url` so query string matches are allowed.

## 0.6.0
initial release.


public/js/editor.js
function AposRedirects() {
  var self = this;
  self.modal = function() {
    self.$el = apos.modalFromTemplate('.apos-redirects-editor', self);
  };

  // Invoked when the modal is ready
  self.init = function(callback) {
    self.$list = self.$el.find('[data-list]');
    self.$template = self.$list.find('[data-item].apos-template');
    self.$template.remove();

    self.$el.on('click', '[data-new]', function() {
      var $item = self.addItem({});
      $item.findByName('from').focus();
      return false;
    });

    self.$el.on('click', '[data-update]', function() {
      var $item = $(this).closest('[data-item]');
      if ($item.length) {
        var from = $item.find('[name="from"]').val();
        if (from.substr(0, 1) !== '/') {
          alert('The first character of the "from" address must be /.');
          return false;
        }
        if (from === '/') {
          alert('To prevent accidental disabling of the entire site, you may not redirect / (the home page).');
          return false;
        }
        // The server treats this as an upsert operation keyed on the "from" URL, so be
        // sure to purge any duplicates browser-side too
        self.$list.find('[data-item]').each(function() {
          if (this === $item[0]) {
            return;
          }
          var $other = $(this);
          if ($other.find('[name="from"]').val() === $item.find('[name="from"]').val()) {
            $other.remove();
          }
        });
        $.jsonCall('/apos-redirects/update', {
          _id: $item.attr('data-id'),
          from: $item.find('[name="from"]').val(),
          to: $item.find('[name="to"]').val(),
          reason: $item.find('[name="reason"]').val()
        }, function(response) {
          if (response.status !== 'ok') {
            alert('An error occurred. Please try again.');
          } else {
            $item.attr('data-id', response.result._id);
            $item.addClass('apos-saving');
            var t = setTimeout(function() {$item.removeClass('apos-saving')}, 1000);
          }
        });
      }
      return false;
    });

    self.$el.on('click', '[data-remove]', function() {
      var $item = $(this).closest('[data-item]');
      if ($item.length) {
        $.jsonCall('/apos-redirects/remove', { _id: $item.attr('data-id') }, function(result) {
          if (result.status !== 'ok') {
            alert('An error occurred. You may have two redirects with the same "from" URL.');
          } else {
            $item.remove();
          }
        });
      }
      return false;
    });

    return self.load(callback);
  };

  self.addItem = function(item) {
    var $newItem = apos.fromTemplate(self.$template);
    self.$list.append($newItem);
    $newItem.attr('data-id', item._id);
    _.each(item, function(val, key) {
      $newItem.findByName(key).val(val);
    });
    return $newItem;
  };

  self.load = function(callback) {
    $.getJSON('/apos-redirects/load', {}, function(response) {
      if (response.status !== 'ok') {
        return alert('An error occurred. Please try again.');
      }
      _.each(response.result, function(item) {
        self.addItem(item);
      });
      return callback(null);
    });
  };

  self.setup = function() {
    $('.apos-redirects-button').click(function() {
      self.modal();
      return false;
    });
  };

  // So it's possible to override self.setup
  apos.afterYield(function() {
    self.setup();
  });
}


index.js
var _ = require('lodash');

module.exports = {
  extend: 'apostrophe-pieces',
  name: 'apostrophe-redirect',
  alias: 'redirects',
  label: 'Redirect',
  pluralLabel: 'Redirects',
  searchable: false,
  adminOnly: true,
  // Default type being joined for internal redirects.
  // Can be overwritten project level with an array of
  // multiple piece types: ex: [ 'apostrophe-page', 'news', 'people' ]
  withType: 'apostrophe-page',
  // Default status code. Must be one of the valid choices
  // for the `statusCode` select field
  statusCode: 302,
  openGraph: false, // Disables apostrophe-open-graph for redirects
  seo: false, // Disables apostrophe-seo for redirects
  sitemap: false, // Disables apostrophe-site-map for redirects
  addFields: [
    {
      name: 'redirectSlug',
      // This is *not* type: 'slug' because we want to let you match any
      // goldang nonsense the old site had in there, including mixed case
      type: 'string',
      label: 'Old URL',
      help: 'Format with leading / such as /old-url',
      required: true
    },
    {
      // contextual flag set so you don't see it in the editor; this gets updated
      // by beforeSave
      type: 'slug',
      name: 'slug',
      label: 'Slug',
      required: true,
      contextual: true
    },
    {
      name: 'title',
      label: 'Description',
      type: 'string',
      required: true
    },
    {
      // contextual: true to hide this property
      type: 'boolean',
      name: 'published',
      label: 'Published',
      required: true,
      def: true,
      contextual: true
    },
    {
      name: 'urlType',
      label: 'Link To',
      type: 'select',
      choices: [
        { label: 'Internal Page', value: 'internal', showFields: [ '_newPage' ] },
        { label: 'External URL', value: 'external', showFields: [ 'externalUrl' ] }
      ]
    },
    {
      name: 'ignoreQueryString',
      label: 'Ignore query string when matching.',
      type: 'boolean',
      def: false
    },
    {
      name: '_newPage',
      type: 'joinByOne',
      withType: 'apostrophe-page',
      label: 'Page Title',
      idField: 'pageId',
      filters: {
        projection: { slug: 1, title: 1, _url: 1 },
        // Admins set up redirects, so it's OK for non-admins to follow them anywhere
        // (they won't actually get access without logging in)
        permission: false
      }
    },
    {
      name: 'externalUrl',
      label: 'URL',
      type: 'url'
    },
    {
      name: 'statusCode',
      label: 'Redirect Type',
      type: 'select',
      help: 'Test new redirects as temporary redirects first. Permanent redirects are an SEO best practice, but only if they are correct.',
      choices: [
        { label: 'Temporary', value: '302' },
        { label: 'Permanent', value: '301' }
      ],
      def: '302'
    }
  ],
  removeFields: ['tags'],
  arrangeFields: [
    {
      name: 'info',
      label: 'Info',
      fields: [
        'redirectSlug',
        'title',
        'urlType',
        'ignoreQueryString',
        '_newPage',
        'externalUrl',
        'statusCode'
      ]
    }
  ],

  beforeConstruct: function (self, options) {
    var _newPage = _.find(options.addFields, { name: '_newPage' });
    _newPage.withType = options.withType;

    var field = _.find(options.addFields, { name: 'statusCode' });
    if (!field) {
      return;
    }
    field.def = options.statusCode.toString();
    if (field.def === '301') {
      field.help = 'By default, redirects are permanent and Google will cache them. Please proofread this carefully or test first with the "temporary" setting.';
    }
  },

  construct: function (self, options) {
    self.beforeSave = function (req, doc, options, callback) {
      // prefix the actual slug so it's not treated as a page
      doc.slug = 'redirect-' + doc.redirectSlug;

      if (!doc.title) {
        doc.title = doc.redirectSlug;
      }

      return callback(null);
    };

    // Check to see if a redirect exists before sending user on their way
    self.expressMiddleware = async (req, res, next) => {
      let slug = req.url;
      let pathOnly = slug.split('?')[0];
      let redirectRegEx = new RegExp(`^redirect-${self.apos.utils.regExpQuote(pathOnly)}(\\?.*)?$`);

      const workflow = self.apos.modules['apostrophe-workflow'];
      const locales = workflow && Object.keys(workflow.locales).filter(locale => !locale.endsWith('-draft'));
      const $or = locales && [
        {
          workflowLocale: {
            $in: locales
          }
        },
        {
          workflowLocale: {
            $exists: 0
          }
        }
      ];

      try {
        const criteria = {
          slug: redirectRegEx,
          type: self.name,
          trash: {
            $ne: true
          },
          published: true
        };
        if ($or) {
          criteria.$or = $or;
        }
        const results = await self.apos.docs.db.find(criteria).project({
          title: 1,
          slug: 1,
          urlType: 1,
          ignoreQueryString: 1,
          pageId: 1,
          type: 1,
          externalUrl: 1,
          redirectSlug: 1,
          statusCode: 1,
          _newPage: 1,
          workflowLocale: 1
        }).toArray();
        const target = results &&
          results.find(result => result.redirectSlug === slug) ||
          results.find(result => result.redirectSlug === pathOnly && result.ignoreQueryString);
        if (!target) {
          return next();
        }
        let status = parseInt(target.statusCode);
        if (isNaN(status) || !status) {
          status = 302;
        }
        if (target.urlType === 'internal') {
          const criteria = {
            _id: target.pageId,
            trash: {
              $ne: true
            },
            published: true
          };
          if ($or) {
            criteria.$or = $or;
          }
          const doc = await self.apos.docs.db.findOne(criteria);
          if (doc) {
            const manager = self.apos.docs.getManager(doc.type);
            if (!manager) {
              return next();
            }
            let oldLocale;
            try {
              oldLocale = req.locale;
              req.locale = doc.workflowLocale;
              target._newPage = await manager.find(req, { _id: target.pageId }).toObject();
            } finally {
              req.locale = oldLocale;
            }
          }
          if (!target._newPage) {
            return next();
          }
          return req.res.redirect(status, target._newPage._url);
        } else if (target.urlType === 'external' && target.externalUrl.length) {
          return req.res.redirect(status, target.externalUrl);
        }
        return next();
      } catch (e) {
        self.apos.utils.error(e);
        return next();
      }
    };
  }
};
